src/RpcServer/RpcServer.SmartContract.cs

```diff
@@ -20,97 +20,106 @@
-              {
-                  var obj = new JObject();
-                  obj["eventname"] = n.EventName;
-                  obj["contract"] = n.ScriptHash.ToString();
-                  obj["state"] = ToJson(n.State, settings.MaxIteratorResultItems);
-                  return obj;
-              }));
-            if (useDiagnostic)
-            {
-                Diagnostic diagnostic = (Diagnostic)engine.Diagnostic;
-                json["diagnostics"] = new JObject()
+            Session session = new(system, script, signers, witnesses, settings.MaxGasInvoke, useDiagnostic ? new Diagnostic() : null);
+            try
+            {
+                json["script"] = Convert.ToBase64String(script);
+                json["state"] = session.Engine.State;
+                json["gasconsumed"] = session.Engine.GasConsumed.ToString();
+                json["exception"] = GetExceptionMessage(session.Engine.FaultException);
+                json["notifications"] = new JArray(session.Engine.Notifications.Select(n =>
                 {
-                    ["invokedcontracts"] = ToJson(diagnostic.InvocationTree.Root),
-                    ["storagechanges"] = ToJson(engine.Snapshot.GetChangeSet())
-                };
+                    var obj = new JObject();
+                    obj["eventname"] = n.EventName;
+                    obj["contract"] = n.ScriptHash.ToString();
+                    obj["state"] = ToJson(n.State, session);
+                    return obj;
+                }));
+                if (useDiagnostic)
+                {
+                    Diagnostic diagnostic = (Diagnostic)session.Engine.Diagnostic;
+                    json["diagnostics"] = new JObject()
+                    {
+                        ["invokedcontracts"] = ToJson(diagnostic.InvocationTree.Root),
+                        ["storagechanges"] = ToJson(session.Engine.Snapshot.GetChangeSet())
+                    };
+                }
+                try
+                {
+                    json["stack"] = new JArray(session.Engine.ResultStack.Select(p => ToJson(p, session)));
+                }
+                catch (InvalidOperationException)
+                {
+                    json["stack"] = "error: invalid operation";
+                }
+                if (session.Engine.State != VMState.FAULT)
+                {
+                    ProcessInvokeWithWallet(json, signers);
+                }
             }
-            try
+            catch
             {
-                json["stack"] = new JArray(engine.ResultStack.Select(p => ToJson(p, settings.MaxIteratorResultItems)));
+                session.Dispose();
+                throw;
             }
-            catch (InvalidOperationException)
+            if (session.Iterators.Count == 0 || !settings.SessionEnabled)
             {
-                json["stack"] = "error: invalid operation";
+                session.Dispose();
             }
```

Comment:
Could you, please, clarify: if sessions [are supposed to be disabled on the public servers](https://github.com/neo-project/neo-modules/pull/715#discussion_r887395316) (mainnet, testnet), then there's no way to get iterator result from `invokefunction` and we should use `invokescript` instead to manually traverse iterator and push its values on stack. 

That's not quite convenient IMO, maybe we still better to keep the old way of iterator dumping if sessions are disabled?